GCP CLI support: server-side request handler

Tener opened the pull request:
Server-side request handler forwarding the requests to GCP, required for https://github.com/gravitational/teleport/issues/17257

To be applied on top of https://github.com/gravitational/teleport/pull/19788.

Tener committed:
GCP CLI support: CA changes

Tener committed:
GCP CLI support: auth changes

Tener committed:
GCP CLI support: API changes

Tener committed:
GCP CLI support: RBAC changes

Tener committed:
GCP CLI support: server-side request handler

GavinFrazar commented on lib/srv/app/gcp/handler_test.go:
```diff
@@ -0,0 +1,134 @@
+			},
+			checkErr: func(t require.TestingT, err error, i ...interface{}) {
+				require.ErrorContains(t, err, "timeout waiting for access token for 5s")
+				require.True(t, errors.Is(err, context.DeadlineExceeded))
```
nit
```suggestion
				require.ErrorIs(t, err, context.DeadlineExceeded))
```

GavinFrazar reviewed (approved):
LGTM

Tener commented on lib/srv/app/azure/handler.go:
```diff
@@ -89,8 +89,13 @@ type handler struct {
 }
 
-// NewAzureHandler creates a new instance of an http.Handler for Azure requests.
+// NewAzureHandler creates a new instance of a http.Handler for Azure requests.
 func NewAzureHandler(ctx context.Context, config HandlerConfig) (http.Handler, error) {
+	return newAzureHandler(ctx, config)
+}
+
+// newAzureHandler creates a new instance of a handler for Azure requests. Used by NewAzureHandler and in tests.
+func newAzureHandler(ctx context.Context, config HandlerConfig) (*handler, error) {
```
`newAzureHandler` returns a private type `*handler` with a private method `getToken()`. This is used for tests. In contrast, `NewAzureHandler` returns `http.Handler`, which hides any private methods.

Tener committed:
Review: use specialized functions in tests.

Tener committed:
Review: perform case-insensitive match, add tests, add docs.

Tener committed:
Review: improve docs.

Tener committed:
Review: simplify test.

smallinsky commented on lib/srv/app/gcp/handler.go:
```diff
@@ -0,0 +1,276 @@
+
+type generateAccessTokenFunc func(ctx context.Context, serviceAccount string, scopes []string) (*credentialspb.GenerateAccessTokenResponse, error)
+
+func (s *HandlerConfig) generateAccessTokenDefaultImpl(ctx context.Context, serviceAccount string, scopes []string) (*credentialspb.GenerateAccessTokenResponse, error) {
```
Can we remove the `DefaultIImpl` suffix 
```suggestion
func (s *HandlerConfig) generateAccessToken(ctx context.Context, serviceAccount string, scopes []string) (*credentialspb.GenerateAccessTokenResponse, error) {
```

Tener replied:
That would result in both field and a method with the same name, which isn't allowed.

Tener replied:
Since I wanted to test `generateAccessToken` (which used to be called `generateAccessTokenDefaultImpl`), I changed the extensibility point used in tests to be the cloud clients. This required a few helper types, but we have better test coverage now.

smallinsky commented on lib/srv/app/gcp/handler.go:
```diff
@@ -0,0 +1,276 @@
+type generateAccessTokenFunc func(ctx context.Context, serviceAccount string, scopes []string) (*credentialspb.GenerateAccessTokenResponse, error)
+
+func (s *HandlerConfig) generateAccessTokenDefaultImpl(ctx context.Context, serviceAccount string, scopes []string) (*credentialspb.GenerateAccessTokenResponse, error) {
+	client, err := s.CloudClients.GetGCPIAMClient(ctx)
```
Don't we need close the client by defer client.Close after access token is generated ?

Tener replied:
We do. I've added the tests to ensure this is done too.

Tener replied:
@smallinsky Actually, turns out that we shouldn't close the client. If we do that it will cause errors when a next request comes, and a new token needs to be generated:
```
2023-01-05T11:00:44+01:00 DEBU [GCP:FWD]   Failed to process request. error:[
ERROR REPORT:
Original Error: *apierror.APIError rpc error: code = Canceled desc = grpc: the client connection is closing
Stack Trace:
	github.com/gravitational/teleport/lib/srv/app/gcp/handler.go:299 github.com/gravitational/teleport/lib/srv/app/gcp.(*handler).generateAccessToken
	github.com/gravitational/teleport/lib/srv/app/gcp/handler.go:267 github.com/gravitational/teleport/lib/srv/app/gcp.(*handler).getToken.func1.1
	github.com/gravitational/teleport/lib/utils/fncache.go:135 github.com/gravitational/teleport/lib/utils.FnCacheGet[...].func1
	github.com/gravitational/teleport/lib/utils/fncache.go:199 github.com/gravitational/teleport/lib/utils.(*FnCache).get.func1
	runtime/asm_arm64.s:1172 runtime.goexit
User Message: rpc error: code = Canceled desc = grpc: the client connection is closing] gcp/handler.go:183
```

smallinsky replied:
Ah, The `s.CloudClients.GetGCPIAMClient(ctx)` returns a global shared client so the client should stay unclosed so my comment is not invalid.

smallinsky commented on lib/srv/app/gcp/handler.go:
```diff
@@ -0,0 +1,276 @@
+		return nil
+	}
+
+	s.Log.Debugf("Processing request, sessionId = %q, gcpServiceAccount = %q", sessionCtx.Identity.RouteToApp.SessionID, sessionCtx.Identity.RouteToApp.GCPServiceAccount)
```
Why this log function is in `replaceAuthHeaders` ? 
this function logs  about general request processing but the scope of this function is related to auth header replacement.

Tener replied:
Agreed, I'll move it.

Tener replied:
Done.

smallinsky commented on lib/srv/app/gcp/handler.go:
```diff
@@ -0,0 +1,276 @@
+// the scope list if fixed for now, but could be either be extended or made customizable
+// if an important use case were to be presented.
+var defaultScopeList = []string{
+	"https://www.googleapis.com/auth/cloud-platform",
+
+	"openid",
+	"https://www.googleapis.com/auth/userinfo.email",
+	"https://www.googleapis.com/auth/appengine.admin",
+	"https://www.googleapis.com/auth/sqlservice.login",
+	"https://www.googleapis.com/auth/compute",
```
Could you add the docs for the GCP scopes that we are using ?

Tener replied:
Done.

smallinsky commented on lib/srv/app/gcp/handler.go:
```diff
@@ -0,0 +1,276 @@
+	}
+	recorder := httplib.NewResponseStatusRecorder(w)
+	s.fwd.ServeHTTP(recorder, fwdRequest)
+	status := uint32(recorder.Status())
```
It looks like in each case of `ecorder.Status` use the return value if casted to uint32.  Can we change the type returned by recorder.Status ?

Tener replied:
You could argue that `ResponseStatusRecorder.Status()` should return int because this is consistent with constants in `http` package. Then again, this type was refactored in https://github.com/gravitational/teleport/pull/19387 by @GavinFrazar, and the `uint32` vs. `int` was discussed therein. Any thoughts on this @GavinFrazar ?

GavinFrazar replied:
I think we could change the type for recorder status, it really shouldn't matter. The http package in go uses `int` for status code, as far as I can tell, for simplicity. They don't ever return a negative integer, but they also don't need the extra range of positive values provided by `uint32`. I will open a PR refactoring this. You can just leave it as-is for now.

Tener replied:
Ok, thanks!

smallinsky commented on lib/srv/app/azure/handler.go:
```diff
@@ -285,7 +290,16 @@ func (s *handler) getToken(ctx context.Context, managedIdentity string, scope st
 	})
+
+	if err != nil {
+		if timeoutCtx.Err() == err {
```
This condition looks wield at first glance. 

Can we just check err.Is(err, context.DeadlineExceeded) ?

Tener replied:
The difference is that we check that _our_ timeout has expired, whereas `err.Is(err, context.DeadlineExceeded)` checks that _some_ timeout has expired.

Tener replied:
After refactoring to use `clock.After` the offending piece of code is gone.

smallinsky commented on lib/srv/app/azure/handler_test.go:
```diff
@@ -0,0 +1,134 @@
+			name: "timeout",
+			config: HandlerConfig{
+				getAccessToken: func(ctx context.Context, managedIdentity string, scope string) (*azcore.AccessToken, error) {
+					time.Sleep(getTokenTimeout * 2)
```
Don't sleep in test -   adding 2 seconds delay in unit test in not good idea.

Tener replied:
`getToken` uses `context.WithTimeout`, which isn't easily testable. I could potentially use https://pkg.go.dev/github.com/jonboulle/clockwork, but I'd have to rewrite it to use `clock.After`. Perhaps I will.

Tener replied:
Done: `clock.After` is used, we no longer have `time.Sleep` in tests.

Tener replied:
Actually, I have only fixed the GCP implementation; the Azure implementation needs a fix too.

Tener replied:
Done for Azure too.

smallinsky reviewed (commented):
First pass.

Tener commented:
@smallinsky 

I have moved this section to https://github.com/gravitational/teleport/pull/19790

> **TODO:** Testing:
> 
> * [ ]  Google Storage interaction with `gsutil` [tool](https://cloud.google.com/storage/docs/gsutil)
> * [ ]  Terraform GCP provider.

Tener committed:
Review: use plural form `GCPServiceAccountsASN1ExtensionOID`

Tener committed:
Review: improve GCP handler.

- make debug log entry in appropriate place
- document the default scope list
- replace `context.WithTimeout` with `s.Clock.After` to avoid `time.Sleep` in tests.
- cloud client is now being mocked, making `generateAccessToken` testeable.
- update tests

Tener committed:
Review: improve Azure handler.

- replace `context.WithTimeout` with `s.Clock.After` to avoid `time.Sleep` in tests.

greedy52 reviewed (approved):
LGTM. 

Only concern is on the pre-defined list of scopes. Not sure how often users need something else.

Tener committed:
Merge branch 'master' into tener/gcloud-cli-handler

Tener committed:
Merge branch 'master' into tener/gcloud-cli-handler

Tener committed:
Merge branch 'master' into tener/gcloud-cli-handler
